### ISSUE TITLE
Create an A/B test for non-recommended install button disclaimer

### ISSUE THREAD

**Issue by bobsilverberg** [mozilla/addons-frontend#8137]
We have discussed the possibility of adding a disclaimer under the install button for non-recommended add-ons, and it was decided we would conduct an A/B test before implementing this for all users.

The final copy for this message (from the Google Doc) is "This extension isn’t monitored by Mozilla. Make sure you trust the extension before you install it. Learn more", where the "Learn more" link will point to the same SUMO article as the one pointed to by the Recommended badge.

As per a conversation with @jvillalobos, we can conduct this test using 50% of AMO users as the cohort who sees the message, while the other 50% will not see the message.

I will discuss any data requirements for the A/B test with my CoE mentor.

**Comment by bobsilverberg** [comment]
@brassy- Would you be able to give me an idea of what this should look like? I believe we want the message below the install button, but should it be wrapped below the button, or extend across the width of the card? Also, what size/color/ etc. should be used?

**Comment by bobsilverberg** [comment]
Here is what I am proposing in terms of data collection for the A/B test:

1. All AMO users will be allocated into one of two cohorts, and this setting will be tracked in a cookie. Members of one cohort will see an install warning for non-recommended add-ons and members of the other will not see an install warning.
2. Whenever an install button for a non-recommended add-on is shown, an event will be sent to GA to indicate which cohort the user is in. Thus, we will be able to see the number of times a user could have been presented with an install warning, broken down by cohort.
3. Whenever an install button for a non-recommended add-on is clicked, an event will be sent to GA to indicate which cohort the user is in. This should give us the ability to determine how often users install non-recommended add-ons, broken down by cohort.

**Comment by willdurand** [comment]
@bobsilverberg  :+1:

**Comment by felixlawrence** [comment]
This sounds very doable from an experimental design point of view. If I understand correctly, your goal is to work out what % of users are dissuaded by the warning and do not install the non-recommended addon. I suspect you would also be interested in knowing, whether users who see the warning for a non-recommended addon are more or less likely to continue browsing AMO, and whether they are more or less likely to go on to download a _recommended_ addon. I presume Google Analytics is already collecting page views, but for this experiment you might also want to send events when _recommended_ addons are shown and clicked.

Regarding point (1), the safest pattern is not to allocate _all_ AMO users, but instead only allocate users as they load a page for a non-recommended addon. In theory the same data can be inferred from your proposed scheme, but in my experience implementation failures are rarer and safer when the allocation and enrollment is done as late as possible.

Were you intending to use Google Analytics Experiments to allocate users to branches, or [Traffic cop](https://github.com/mozilla/trafficcop/), or something else?

There is going to be a bit of a learning curve for me on this experiment, because this is in the race to be my first web-based experiment at Mozilla.

**Comment by bobsilverberg** [comment]
> This sounds very doable from an experimental design point of view. If I understand correctly, your goal is to work out what % of users are dissuaded by the warning and do not install the non-recommended addon. I suspect you would also be interested in knowing, whether users who see the warning for a non-recommended addon are more or less likely to continue browsing AMO, and whether they are more or less likely to go on to download a _recommended_ addon. I presume Google Analytics is already collecting page views, but for this experiment you might also want to send events when _recommended_ addons are shown and clicked.

This is a good point, thanks @felixlawrence. In order to track all of the behaviour of users in each cohort, I think we should introduce a custom dimension which tracks the cohort a user belongs to. This will allow us to look at any existing GA stats (like the time spent on the site) broken down by cohort. Do you agree that this seems like a good idea?
 
> 
> Regarding point (1), the safest pattern is not to allocate _all_ AMO users, but instead only allocate users as they load a page for a non-recommended addon. In theory the same data can be inferred from your proposed scheme, but in my experience implementation failures are rarer and safer when the allocation and enrollment is done as late as possible.

Thanks for this advice. I can set it up so that the user is only allocated into a cohort once they view an install button for a non-recommended add-on.

> 
> Were you intending to use Google Analytics Experiments to allocate users to branches, or [Traffic cop](https://github.com/mozilla/trafficcop/), or something else?

We are using logic on the AMO site to allocate users and store the users' cohort in a cookie to that a user will remain in a cohort until that cookie is cleared. We just use a basic random number to place half of the users in each cohort.

> 
> There is going to be a bit of a learning curve for me on this experiment, because this is in the race to be my first web-based experiment at Mozilla.

I look forward to working with you on this!

**Comment by felixlawrence** [comment]
> In order to track all of the behaviour of users in each cohort, I think we should introduce a custom dimension which tracks the cohort a user belongs to. This will allow us to look at any existing GA stats (like the time spent on the site) broken down by cohort. Do you agree that this seems like a good idea?

I have little hands-on experience with GA (none at Mozilla), so I'm not in a good position to assess whether this is the most appropriate mechanism to achieve this goal, but it sounds reasonable. Let me know if you're unsure about this to the point that you want me to research this point.

> We are using logic on the AMO site to allocate users and store the users' cohort in a cookie to that a user will remain in a cohort until that cookie is cleared. We just use a basic random number to place half of the users in each cohort.

Is this an established piece of functionality? It would make data scientists' lives easier if we re-used components (like Traffic cop) where possible. Less code to review, and a smaller set of quirks to learn. But this is a relatively low priority for this test because I don't know enough to confidently recommend any solution as having the least quirks :)

> I look forward to working with you on this!
Yeah, it should be fun!

**Comment by bobsilverberg** [comment]
> > In order to track all of the behaviour of users in each cohort, I think we should introduce a custom dimension which tracks the cohort a user belongs to. This will allow us to look at any existing GA stats (like the time spent on the site) broken down by cohort. Do you agree that this seems like a good idea?
> 
> I have little hands-on experience with GA (none at Mozilla), so I'm not in a good position to assess whether this is the most appropriate mechanism to achieve this goal, but it sounds reasonable. Let me know if you're unsure about this to the point that you want me to research this point.

I think we're okay going forward with this. We did this in another experiment I was involved with. On the other hand, it might be worth just passing this by one of you colleagues who might be more familiar with GA just to double check. I don't want to ask you to do a lot of research into it, but a quick check could be a good idea.

> > We are using logic on the AMO site to allocate users and store the users' cohort in a cookie to that a user will remain in a cohort until that cookie is cleared. We just use a basic random number to place half of the users in each cohort.
> 
> Is this an established piece of functionality? It would make data scientists' lives easier if we re-used components (like Traffic cop) where possible. Less code to review, and a smaller set of quirks to learn. But this is a relatively low priority for this test because I don't know enough to confidently recommend any solution as having the least quirks :)

Yes, this is an established piece of functionality that we've used for a number of A/B tests in the past, so there's not really any code that will need to be reviewed for this.

> > I look forward to working with you on this!
> 
> Yeah, it should be fun!

👍

**Comment by bobsilverberg** [comment]
> > Regarding point (1), the safest pattern is not to allocate _all_ AMO users, but instead only allocate users as they load a page for a non-recommended addon. In theory the same data can be inferred from your proposed scheme, but in my experience implementation failures are rarer and safer when the allocation and enrollment is done as late as possible.
> 
> Thanks for this advice. I can set it up so that the user is only allocated into a cohort once they view an install button for a non-recommended add-on.

I took a look at our code for this, and it's not currently designed to support this. It currently would allocate a user into a cohort as soon as (but not before) they view an install button for *any* add-on, not just for a non-recommended add-on. Do you think that's a reasonable compromise?

**Comment by felixlawrence** [comment]
> > > Regarding point (1), the safest pattern is not to allocate _all_ AMO users, but instead only allocate users as they load a page for a non-recommended addon. In theory the same data can be inferred from your proposed scheme, but in my experience implementation failures are rarer and safer when the allocation and enrollment is done as late as possible.
> > 
> > 
> > Thanks for this advice. I can set it up so that the user is only allocated into a cohort once they view an install button for a non-recommended add-on.
> 
> I took a look at our code for this, and it's not currently designed to support this. It currently would allocate a user into a cohort as soon as (but not before) they view an install button for _any_ add-on, not just for a non-recommended add-on. Do you think that's a reasonable compromise?

It won't _invalidate_ the experiment but it will add noise. What % of sessions that involve viewing any addon involve a view of a non-recommended add-on? If it's >50% then it's no problem, if it's less than say 5% then we'll need to do some mitigation.

**Comment by bobsilverberg** [comment]
> > I took a look at our code for this, and it's not currently designed to support this. It currently would allocate a user into a cohort as soon as (but not before) they view an install button for _any_ add-on, not just for a non-recommended add-on. Do you think that's a reasonable compromise?
> 
> It won't _invalidate_ the experiment but it will add noise. What % of sessions that involve viewing any addon involve a view of a non-recommended add-on? If it's >50% then it's no problem, if it's less than say 5% then we'll need to do some mitigation.

I would venture to say that it's at least 50%, possibly a lot more. What do you think @jvillalobos || @devaneymoz?  What % of sessions that involve viewing any add-on involve a view of a non-recommended add-on?

**Comment by jvillalobos** [comment]
Considering that the most popular extension (AdBlock Plus) is not Recommended, I think it's pretty much impossible for the traffic to non-Recommended to drop anywhere near 5%. The Recommended list is fairly small (under 200), so I think it's safe to say that it will still get a minority of the traffic. Getting it to 50/50 would definitely be a success, but I don't expect that to happen so quickly, if at all.

**Comment by ioanarusiczki** [comment]
@bobsilverberg Hi Bob, here are my testing results with FF68(Win10) on AMO dev:

**the  enrollment event into withExperiment**

- the GA event "AMO Experiment Enrollment - installButtonWarning" is only shown once per profile. If I cleared the cookie the cohort may change but this type of event won't be sent again.

**the InstallWarning component**

- the yellow warning message is displayed for `includeWarning` cohort on the add-ons detail page, version page and guide lines. Clicking Learn more will redirect to the SUMO article as the one pointed by the recommended badge.
- the yellow warning message for `includeWarning` cohort is available only for extensions. An add-on which is installed won't display the message on the detail page. Only after `remove` button is clicked the message will be available on the detail pages, version page or guide lines.
- I did not see the message being available for the `excludeWarning` cohort.
-"AMO Install Button Warning Experiment - Display" GA event is sent whenever an install button in the "+Add to Firefox" state is displayed for a non-recommended add-on. 

**GA events sent when install button is clicked**  
- work as described [above](https://github.com/mozilla/addons-frontend/issues/8137#issue-452999111). Also on AMO stage the experiment is not enabled and the Browser Console shows `notInExperiment` value. Also, on AMO stage there is no cookie set or yellow messages displayed (my guess, it's as it should). 

**The 50-50 cohorts test result**
- Test one:  I created 10 new profiles and I simply opened AMO dev and landed on a detail page, had 6 times `excludeWarning` and 4 times `includeWarning`. 
- Test two: I used one profile, loaded AMO dev and landed on a detail page then deleted data from the Options of the Browser, and refreshed the detail page of an add-on while watching the cookie from Storage. I got 5 times  `excludeWarning` and 5 times `includeWarning`. 
- Test three: A profile which initially gave the `excludeWarning` value to the cookie closed and reopend (without clearing data). The third time reopened I've noticed that it changed the value to `includeWarning`. After that I closed and restarted the profile 10 times but the `includeWarning` value did not change into `excludeWarning`. 

My question mark is about this last test - why once `includeWarning` value is set it won't change to `excludeWarning` because the Storage shows that the cookie expires per Session. (still I find it awkard if it should behave like showing then not showing then again showing the message every time the user closes or reopens the browser).
On the other hand an [above comment](https://github.com/mozilla/addons-frontend/issues/8137#issuecomment-512812369) says that the cookie must be cleared so that the cohort is changed. 

But as a general behavior, those 6 profiles that initially had the `excludeWarning` value received the `includeWarning` at one point , without me clearing the cookie but by simply reopening the profile, and  no `includeWarning` profile changed to `excludeWarning` without clearing manually the cookie . 

And the other question: when is the experiment enabled on stage and prod ?

**Comment by ioanarusiczki** [comment]
@bobsilverberg 
- I re-checked the enrollment event and I confirm it is behaving correctly. My testing method was clearing the cookie and then refresh an add-ons detail page. This way the Console did not show the initial enrollment event. I tried to be on AMO somewhere in a search result or homepage and then click the detail page and this way the enrollment event shows every time after I cleared the cookie. 
-  I can confirm that the corresponding cohort is associated with the event.
If I have  `installButtonWarningExperiment:"excludeWarning"`
Console will display 
`[GA: ON] sendEvent {"hitType":"event","eventCategory":"AMO Install Button Warning Experiment - Click-not_recommended","eventAction":"excludeWarning","eventLabel":"Brief"}`
and 
 `[GA: ON] sendEvent {"hitType":"event","eventCategory":"AMO Install Button Warning Experiment - Display","eventAction":"excludeWarning","eventLabel":"Brief"}`

The same for `includeWarning`. 

- I agree with the cookie persisting across sessions.

- I ran into a weird behavior today on a profile that initially got the `excludeWarning` and remained on `excludeWarning` no matter how many times I closed or reopened the browser (and the only profile which did not change the cookie to `include` so far) I tried to reproduce it on other profiles without any success but I should try further test it the following days. Also, I could not reproduce it with profiles that have `includeWarning` cookie and add-ons installed - none display the warning message in addon:manager. 
So, on this profile with `excludeWarning` I installed add-ons and today when I reopened it was still on `excludeWarning` but in the addon:manager two of the installed add-ons display the yellow warning message. It's as weird as it sounds. Please take a look:

![User excluded from cohort2](https://user-images.githubusercontent.com/33448286/63934157-636a6180-ca63-11e9-8ea5-b4f71ee3a798.gif)

**Comment by ioanarusiczki** [comment]
@bobsilverberg Bottom line, installing needs further investigations. I'll try some more tests tomorrow and on Monday.

**Comment by bobsilverberg** [comment]
> @bobsilverberg
> 
> * I re-checked the enrollment event and I confirm it is behaving correctly. My testing method was clearing the cookie and then refresh an add-ons detail page. This way the Console did not show the initial enrollment event. I tried to be on AMO somewhere in a search result or homepage and then click the detail page and this way the enrollment event shows every time after I cleared the cookie.

Great, thanks!

> 
> * I can confirm that the corresponding cohort is associated with the event.
>       If I have  `installButtonWarningExperiment:"excludeWarning"`
>       Console will display
>       `[GA: ON] sendEvent {"hitType":"event","eventCategory":"AMO Install Button Warning Experiment - Click-not_recommended","eventAction":"excludeWarning","eventLabel":"Brief"}`
>       and
>       `[GA: ON] sendEvent {"hitType":"event","eventCategory":"AMO Install Button Warning Experiment - Display","eventAction":"excludeWarning","eventLabel":"Brief"}`
> 
> 
> The same for `includeWarning`.
>

Also thanks!
 
> * I agree with the cookie persisting across sessions.
>

This should be fixed via #8515 
 
> * I ran into a weird behavior today on a profile that initially got the `excludeWarning` and remained on `excludeWarning` no matter how many times I closed or reopened the browser (and the only profile which did not change the cookie to `include` so far) I tried to reproduce it on other profiles without any success but I should try further test it the following days. 

This could just be an unexpected randomness. I don't see anything in the code that would cause a cookie to persist a particular value, even after being cleared. It might just be bad luck that the same value was assigned a number of times in a row.

> Also, I could not reproduce it with profiles that have `includeWarning` cookie and add-ons installed - none display the warning message in about:addons.
> So, on this profile with `excludeWarning` I installed add-ons and today when I reopened it was still on `excludeWarning` but in the about:addons two of the installed add-ons display the yellow warning message. It's as weird as it sounds. Please take a look:
> 
> 
> ![User excluded from cohort2](https://user-images.githubusercontent.com/33448286/63934157-636a6180-ca63-11e9-8ea5-b4f71ee3a798.gif)

None of the code in addons-frontend controls anything in about:addons, so this is expected. That message has nothing to do with this message/experiment.

**Comment by ioanarusiczki** [comment]
@bobsilverberg The message I was mentioning [above](https://github.com/mozilla/addons-frontend/issues/8137#issuecomment-526132144) which appears randomly on profiles I've tested with  is not related to the experiment. 
I realized today that the text is different and the link redirects to a different SUMO article. They have a big similarity because of the style used. 
Explanation would be a pref `xpinstall.signatures.required` set on false for the FF unbranded version I've tested with, in order to check on installed add-ons on dev/stage environments. 

This is today on stage - where the experiment is not enabled yet 

![on stage](https://user-images.githubusercontent.com/33448286/64116076-f1af5200-cd99-11e9-94c3-3e14c230377d.png)

**Comment by bobsilverberg** [comment]
> 
> This is today on stage - where the experiment is not enabled yet
> 
> ![on stage](https://user-images.githubusercontent.com/33448286/64116076-f1af5200-cd99-11e9-94c3-3e14c230377d.png)
>

It sounds like you are already aware of this, but the message above has nothing to do with this patch. That is the about:addons page, which in fact isn't controlled by the code in addons-frontend at all.
